src/Driver/PgSQL/Connection.php

```diff
@@ -0,0 +1,130 @@
+
+namespace Doctrine\DBAL\Driver\PgSQL;
+
+use Doctrine\DBAL\Driver\ServerInfoAwareConnection;
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\SQL\Parser;
+use Doctrine\Deprecations\Deprecation;
+use PgSql\Connection as PgSqlConnection;
+use TypeError;
+
+use function assert;
+use function get_class;
+use function gettype;
+use function is_object;
+use function is_resource;
+use function pg_escape_string;
+use function pg_get_result;
+use function pg_result_error;
+use function pg_send_prepare;
+use function pg_version;
+use function sprintf;
+use function uniqid;
+
+final class Connection implements ServerInfoAwareConnection
+{
+    /** @var PgSqlConnection|resource */
+    private $connection;
+
+    private Parser $parser;
+
+    /** @param PgSqlConnection|resource $connection */
+    public function __construct($connection)
+    {
+        if (! is_resource($connection) && ! $connection instanceof PgSqlConnection) {
+            throw new TypeError(sprintf(
+                'Expected connection to be a resource of an instance of %s, got %s.',
+                PgSqlConnection::class,
+                is_object($connection) ? get_class($connection) : gettype($connection),
+            ));
+        }
+
+        $this->connection = $connection;
+        $this->parser     = new Parser(false);
+    }
+
+    public function prepare(string $sql): Statement
+    {
+        $visitor = new ConvertParameters();
+        $this->parser->parse($sql, $visitor);
+
+        $statementName = uniqid('dbal', true);
+        $success       = (bool) pg_send_prepare($this->connection, $statementName, $visitor->getSQL());
+
+        assert($success);
+
+        $result = @pg_get_result($this->connection);
+        assert($result !== false);
+
+        if ((bool) pg_result_error($result)) {
+            throw Exception::fromResult($result);
+        }
+
+        return new Statement($this->connection, $statementName, $visitor->getParameterMap());
+    }
+
+    public function query(string $sql): Result
+    {
+        return $this->prepare($sql)->execute();
+    }
+
+    /** @inheritdoc */
+    public function quote($value, $type = ParameterType::STRING)
+    {
+        return sprintf("'%s'", pg_escape_string($this->connection, $value));
+    }
+
+    public function exec(string $sql): int
+    {
+        return $this->prepare($sql)->execute()->rowCount();
+    }
+
+    /** @inheritdoc */
+    public function lastInsertId($name = null)
+    {
+        if ($name !== null) {
+            Deprecation::triggerIfCalledFromOutside(
+                'doctrine/dbal',
+                'https://github.com/doctrine/dbal/issues/4687',
+                'The usage of Connection::lastInsertId() with a sequence name is deprecated.',
+            );
+
+            return $this->query(sprintf('SELECT CURRVAL(%s)', $this->quote($name)))->fetchOne();
+        }
+
+        return $this->query('SELECT LASTVAL()')->fetchOne();
+    }
+
+    public function beginTransaction(): bool
```

Comment:
Should these be documented with `@return true`?

Reply:
We could do this for most of the non-PRO drivers if we want. On 4.0, we've switched all of them to void.